### DISCUSSION TITLE
Does prisma.$transaction lock rows/always perform asynchronously?

### DISCUSSION THREAD

**Discussion by benschool** [prisma/prisma#19714] [Q&A]
I am using Prisma and MongoDB. I am working on preventing a case where if a function is run twice at exactly the same time, my constraints are ignored. Here is an example of my case:
```
async function changeBalance(userId, amount) {
  await prisma.$transaction(async (tx) => {
    const userBalance = await prisma.user.findFirst({ where: { id: userId }, select: { balance: true }})
    if (userBalance.balance + amount < 0) { throw new Error("This would result in a negative balance") }
    await tx.user.update({
      where: { id: userId },
      data: { balance: userBalance + amount },
    }) 
  })
}
```

If this function is run twice at the exact same time, will the second transaction always wait for the first one?

**Comment by nurul3101** [comment]
Hello @benschool 👋 

MongoDB does not lock the documents during the transaction. Instead, when the transaction is committed, if MongoDB detects that any of the documents involved in the transaction have been modified by another operation during the transaction, it will abort the transaction with a write conflict error. This is known as optimistic concurrency control, you can read more about it in our documentation here: [Optimistic Concurrency Control](https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide#optimistic-concurrency-control)

Hope this helps! Please let us know in case you have any queries. If you don't have any queries, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved.

**Reply by nurul3101** [reply]
I am checking to see if you had a chance to look at my last response. Please let us know in case you need any assistance.

**Reply by nurul3101** [reply]
@benschool Do let us know in case you need any assistance.

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have a similar problem 🙏
